### DISCUSSION TITLE
how to auto login keycloak angular application in embedded browser if user already login

### DISCUSSION THREAD

**Discussion by wanihemant** [keycloak/keycloak#23241] [Q&A]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

authentication

### Describe the bug

We have desktop application and in that application we have used embedded browser for showing some project modules. And we have developed our front end project in angular. So, issue is if user is already logged in desktop application and open that embedded browser then embedded browser showing keycloak login page. we need to skip login page if user is already logged in and redirect to respective embedded browser page. 

### Version

22.0.1

### Expected behavior

 we need to skip login page if user is already logged in and redirect to respective embedded browser page. 

### Actual behavior

If user already logged in still showing login page in embedded browser.

### How to Reproduce?

NA

```[tasklist]
### Tasks
```

**Comment by jonkoops** [comment]
This is a very specific question with multiple layers of abstractions and without a reproducible example, it is therefore likely to never be answered. Also, this is not a bug or feature request, so I am converting this to a discussion.

**Comment by Satishshelake** [comment]
Hello,
   We have access token and trying to explore angular url based on access token. We are struggling to init keycloak service using access token. 
angular application url  will be accessible and init keycloak service in embedded browser based on access token.
We have searched a lot on it. but didn't find examples or references.
Could you suggest approach for above scenario?

**Comment by wanihemant** [comment]
Hi,

I have tried onLoad check-sso option but still it's not working.
Please find the code below

export class KeycloakSecurityService {
  public kc: KeycloakInstance;

  constructor(
    private router: ActivatedRoute,
    private http: HttpClient
  ) { }

  async init() {
    this.kc = new keycloak({
      url: environment?.keycloakRootUrl,
      realm: 'angularRelam',
      clientId: 'client',
    });
    await this.kc.init({
      onLoad: 'check-sso',
    });
  }
}

**Comment by justin-tay** [comment]
The Single-Sign On works by using cookies. The best practice as documented in [RFC 8252 - OAuth 2.0 for Native Apps](https://datatracker.ietf.org/doc/html/rfc8252) is to always use the system browser to query the authorization endpoint. If the different applications aren't sharing cookies you should expect that they need to login to multiple applications.

[RFC 8252 - OAuth 2.0 for Native Apps](https://datatracker.ietf.org/doc/html/rfc8252) has the following section https://datatracker.ietf.org/doc/html/rfc8252#section-8.12 that states

> This best current  practice requires that native apps MUST NOT use embedded user-agents to perform authorization requests and allows that authorization endpoints MAY take steps to detect and block authorization requests  in embedded user-agents.

> Aside from the security concerns, embedded user-agents do not share the authentication state with other apps or the browser, requiring the user to log in for every authorization request, which is often considered an inferior user experience.

**Comment by Satishshelake** [comment]
Thank you for reply.  Our requirement is very custom. We have windows-based UI application. which contains embedded browser to connect some modules in Angular web-based application.  Now Angular application is sso enabled with keycloak angular Library. 
We don't have sso on windows application, it has username & password. we can connect to keycloak server from Windows based application & get access token. 
Can We use above token to establish session in angular application which is running inside embedded browser?

**Reply by justin-tay** [reply]
For SSO to work the browser needs to share the `KEYCLOAK_IDENTITY` cookie. This is typically done by using the same system browser to do the logins as stated in https://datatracker.ietf.org/doc/html/rfc8252.

**Reply by hamza2727** [reply]
@Satishshelake I am also stuck in the same type of issue.
 The requirement is that if the user is logged in on a Windows machine and when he wants to log in to the web application, then the keycloak login page should not come up. The authentication should happen on Windows credentials and allow the user to log in and generate a token using the keycloak
 
 have you resolved the issue?

**Comment by wanihemant** [comment]
Hello

I have set KEYCLOAK_IDENTITY cookies with ID_TOKEN for embedded browser but still it's redirect to keycloak login page.
Do we need to add any other cookies too?

<img width="797" alt="image" src="https://github.com/keycloak/keycloak/assets/144450321/fb4033e3-9dcf-426b-9e85-239e820d3c72">

**Reply by justin-tay** [reply]
The `KEYCLOAK_IDENTITY` cookie value is not the `id_token` value.

**Comment by Satishshelake** [comment]
Thank you for reply. I am new in keycloak authentication flow. I have grand type password direct access flow, which field will be match to KEYCLOAK IDENTITY?

**Reply by justin-tay** [reply]
None of the fields will match. It is an internal session credential that is set on the cookie on authentication.

**Reply by Satishshelake** [reply]
Then It's not possible to fetch information rgarding sso programatically & use that data to set cookies?

**Reply by Satishshelake** [reply]
Could we fetch information like KEYCLOAK IDENTITY regarding sso programmatically? And use that use cookies to launch embedded browser?

**Comment by Satishshelake** [comment]
Could we fetch information like KEYCLOAK IDENTITY regarding sso programmatically? And use that use cookies to launch embedded browser?
